### ISSUE TITLE
Migrations with hot code upgrade

### ISSUE THREAD

**Issue by orikremer** [boldpoker/edeliver#102]
I'm trying to set up a continuously deployed build for a phoenix app using edeliver.
After the initial release (which I did manually) my build has the following step:
```
mix edeliver upgrade production
```

According to the readme the steps taken are:
```
1. Detect current version on all running nodes
2. Validate that all nodes run the same version
3. Build new upgrade from that version to the current version
4. Auto-patch the relup file
5. Deploy (hot code) upgrade while nodes are running
6. Validate that all nodes run the upgraded version
7. Deploy the release to not running nodes
```

If I understand correctly, step 5 does both deploying/copying the new version to the production hosts and also hot upgrading the running version.
I would like to run migrations between those two operations. Is that possible?

**Comment by bharendt** [comment]
Yes, this is possible and you have mainly two options:

 1) Run `edeliver migrate` in a `post_upgrade_release` [post hook](https://github.com/boldpoker/edeliver/wiki/Run-additional-build-tasks)
 2) [(Automaticaly) modify the relup instructions](https://hexdocs.pm/edeliver/1.2.9/Edeliver.Relup.Modification.html) to run the migrations while the hot code upgrade is installed

Because the migrations are included in the upgrade package which is installed, running the migrations can only be done during step 4 and 5 (option 1) or after step 5 (option 2) of the automatic upgrade. 
Option 1) is more easy to implement while option 2) provides more fine-grained control over the upgrade, e.g. allows to suspend processes which are using the database while executing the migrations.

Using a [post hook](https://github.com/boldpoker/edeliver/wiki/Run-additional-build-tasks) could be done like this:

```bash
# ./deliver/config

post_upgrade_release() {
  set_deploy_hosts
  export NODE_ACTION="migrate"
  source $(dirname ${BASH_SOURCE[0]})/../deps/edeliver/strategies/erlang-node-execute
  run
}
```

Using an [automatic relup modification](https://hexdocs.pm/edeliver/1.2.9/Edeliver.Relup.Modification.html) could be done like this:

```elixir
# instruction which runs the migrations while the upgrade is installed
defmodule Edeliver.Relup.Instructions.RunMigrations do
  use Edeliver.Relup.RunnableInstruction

  def insert_where, do: &append_after_point_of_no_return/2

  # the return value is passed to the `run/1` function during the upgrade
  def arguments(_instructions = %Instructions{up_version: up_version}, _config = %Config{name: name}) do
    {name, up_version}
  end

  def run({otp_application_name, up_version}) do
    info "Installing migrations for release version #{up_version}..."
    Edeliver.migrate(otp_application_name, up_version, _ecto_repository = :detect, _direction = :up)
    info "Installed migrations for release version #{up_version}."
  end
end

# automatic relup modification which adds the instruction to the relup file
defmodule Edeliver.Relup.UpgradeModification do
  use Edeliver.Relup.Modification

  def modify_relup(instructions = %Instructions{}, config = %Config{}) do
    # use default modifications for phoenix apps (if you are using a phoenix app)
    # and add custom migration instruction.
    # maybe add other instructions to suspend processes which are using the database 
    # beforethe `RunMigrations` instruction is executed and and resume them afterwards.
    instructions |> Edeliver.Relup.PhoenixModification.modify_relup(config) 
                 |> Edeliver.Relup.Instructions.RunMigrations.modify_relup(config) 
  end

  # use this as default modification my using priority higher than `priority_default`
  def priority, do: priority_user
end
```

**Comment by orikremer** [comment]
Thank you for the detailed explanation.
One thing that wasn't clear to me is on the exact timing of when the migrations will be run.
Will both options run the migrations before the new code is running/accessible?

**Comment by orikremer** [comment]
@bharendt It looks like in order to run migrations before the new code is out I would want to take option 2 (automatic relup).
Any thoughts on adding Edeliver.Relup.Instructions.RunMigrations?

**Comment by bharendt** [comment]
Yes, `post_upgrade_release` is called after the hot code upgrade. If you want to run the migrations before the new code is loaded you need to modify the relup instructions as described in option 2. If you use
```elixir
def insert_where, do: &append_after_point_of_no_return/2
```
to insert the migration instruction into the relup file, it will be executed [before the new code is loaded](https://hexdocs.pm/edeliver/1.2.9/Edeliver.Relup.InsertInstruction.html#append_after_point_of_no_return/2). 
If you want to execute it as first instruction at all, you can use  [`insert_after_point_of_no_return`](https://hexdocs.pm/edeliver/1.2.9/Edeliver.Relup.InsertInstruction.html#insert_after_point_of_no_return/2) instead.

**Comment by onnoowl** [comment]
Hi, I'm trying to accomplish migrations during hot code upgrading too, but neither of these solutions worked for me. The strange thing is that neither of these solutions gave me any errors. I didn't see any errors or migration attempts when upgrading with --verbose, nor did I see any migration errors or any attempt at migration in the server's log. The upgrade just happily completes itself, but no migration attempt takes place. I have to manually do `mix edeliver migrate`.

For option 2, are there any extra steps required to make this work other than just having this code in the codebase?

For both options, are there any changes that need to be made to this code now to get it to work? It has been over a year since it was posted. At this time, is there now a better way to perform migrations during a hot code upgrade? What's the best practice?

**Comment by lessless** [comment]
@onnoowl did you mange to solve it?

**Comment by onnoowl** [comment]
Thanks for the mention @lessless, and helping get that merged in, looks great!

@bharendt, I had a quick question for you. The above relup modification looks like it does exactly what I need it to do, but what about the "--run-migrations" option? I see it mentioned in the README, but I couldn't find any more information on it. How does using --run-migrations differ from using this relup modification approach?

**Comment by bharendt** [comment]
@onnoowl Sorry,  the`--run-migrations` option mentioned in the README is not yet implemented. The idea behind that option is to automatically apply a `RunMigrations` relup modification as above, e.g. in the `DefaultModification` without the need to create a custom relup modification for that.

**Comment by onnoowl** [comment]
Ok, thanks for the information!
